tree/tree/src/TTree.cxx

```diff
@@ -4457,97 +4454,86 @@ Int_t TTree::Fill()
+
+   bool autoFlush = fAutoFlush != 0;
+   bool autoSave  = fAutoSave  != 0;
+
+   if (autoFlush || autoSave) {
+      if (fFlushedBytes > 0) {
+         auto flushEntries = fNClusterRange == 0 ? fEntries : fEntries - fClusterRangeEnd[fNClusterRange-1];
+         autoFlush = flushEntries % fAutoFlush == 0;
+         autoSave  = fEntries % fAutoSave == 0;
+      } else {
          Long64_t zipBytes = GetZipBytes();
-         if ((fAutoFlush<0 && zipBytes > -fAutoFlush)  ||
-             (fAutoSave <0 && zipBytes > -fAutoSave )  ||
-             (fAutoFlush>0 && fEntries%TMath::Max((Long64_t)1,fAutoFlush) == 0) ||
-             (fAutoSave >0 && fEntries%TMath::Max((Long64_t)1,fAutoSave)  == 0) ) {
 
-            //First call FlushBasket to make sure that fTotBytes is up to date.
+         autoFlush = fAutoFlush > 0 ? fEntries == fAutoFlush : zipBytes > -fAutoFlush;
+         autoSave  = fAutoSave  > 0 ? fEntries == fAutoSave  : zipBytes > -fAutoSave;
+
+         // First time that AutoFlush/AutoSave are being called
+         if (autoFlush || autoSave) {
             FlushBaskets();
-            OptimizeBaskets(GetTotBytes(),1,"");
-            if (gDebug > 0) Info("TTree::Fill","OptimizeBaskets called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
+            OptimizeBaskets(GetTotBytes(), 1, "");
+            autoFlush = false; // already flushed, do not flush again below
+
+            fAutoFlush = fEntries; // switch to test on number of entries
+
+            if (gDebug > 0)
+               Info("TTree::Fill", "OptimizeBaskets called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",
+                    fEntries, GetZipBytes(), fFlushedBytes);
+
             fFlushedBytes = GetZipBytes();
-            fAutoFlush    = fEntries;  // Use test on entries rather than bytes
-
-            // subsequently in run
-            if (fAutoSave < 0) {
-               // Set fAutoSave to the largest integer multiple of
-               // fAutoFlush events such that fAutoSave*fFlushedBytes
-               // < (minus the input value of fAutoSave)
-               Long64_t totBytes = GetTotBytes();
-               if (zipBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/zipBytes)/fEntries));
-               } else if (totBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/totBytes)/fEntries));
-               } else {
-                  TBufferFile b(TBuffer::kWrite, 10000);
-                  TTree::Class()->WriteBuffer(b, (TTree*) this);
-                  Long64_t total = b.Length();
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/total)/fEntries));
-               }
-            } else if(fAutoSave > 0) {
-               fAutoSave = fAutoFlush*(fAutoSave/fAutoFlush);
-            }
-            if (fAutoSave!=0 && fEntries >= fAutoSave) AutoSave();    // FlushBaskets not called in AutoSave
```

Comment:
this line seems to be replaced in spirit by new line 4481
```
autoSave  = fAutoSave  > 0 ? fEntries == fAutoSave  : zipBytes > -fAutoSave;
```
Is the result really the same if the TTree is not compressed?

Reply:
Yes, this was a possible problem before. I believe the new branch has this fixed.